### ISSUE TITLE
Update addToDefinition when adding layers to Feature Service or View to be async

### ISSUE THREAD

**Issue by chris-fox** [Esri/solution.js#574]
When adding many layers to a service or view during deployment the addToDefinition call can timeout causing it to fail to deploy. In the python api we changed this logic to chunk the request in batches of 20. From the development team their recommendation is we make an async call to addToDefinition which doesn't have a timeout. With the async request you will get a status url in the response that you can check periodically to see when the operation completes. This seems like the safest way to guarantee all services small or large can be created successfully.

FYI...@MikeMillerGIS

**Comment by MikeMillerGIS** [comment]
email trail for info
TLDR
Use Async and Chunking:

-----
ASync call is the right way. You can always batch with async call.
We have a limit of 64K message size inside the queue for the async job.

So I would think you can always batch like 10-20 layers in one async call if this works reliably.

Khaled

From: Chris Fox <chris_fox@esri.com> 
Sent: Thursday, January 7, 2021 7:02 AM
To: Khaled Hassen <KHassen@esri.com>; Paul Barker <PBarker@esri.com>; Michael Miller <mmiller@esri.com>
Cc: Chris Whitmore <cwhitmore@esri.com>; Jeremy Bartley <jbartley@esri.com>
Subject: RE: Time out on Add To Definition

The larger issue is for our api to clone/deploy feature services as part of Solutions. Some solutions may have feature services with many layers like below. Our question would be is it more reliable for us to do this with an async call or in batches of layers with a synchronous or async call.

-Chris

From: Khaled Hassen <KHassen@esri.com> 
Sent: Thursday, January 7, 2021 9:58 AM
To: Paul Barker <PBarker@esri.com>; Michael Miller <mmiller@esri.com>
Cc: Chris Fox <chris_fox@esri.com>; Chris Whitmore <cwhitmore@esri.com>; Jeremy Bartley <jbartley@esri.com>
Subject: RE: Time out on Add To Definition

Yes we do have a limit in the size of the message in the queue. Not sure if it is the same issue.
But it does seem we need to use async call anyway to avoid time-out.

Michael: can we try the admin API directly and see if this fails as well. You can always add layers to the view in more than one call, each have subset of the layers.

Khaled

From: Paul Barker <PBarker@esri.com> 
Sent: Thursday, January 7, 2021 6:53 AM
To: Khaled Hassen <KHassen@esri.com>; Michael Miller <mmiller@esri.com>
Cc: Chris Fox <chris_fox@esri.com>; Chris Whitmore <cwhitmore@esri.com>; Jeremy Bartley <jbartley@esri.com>
Subject: RE: Time out on Add To Definition

Khaled,  we had changed that in the UI awhile ago but had to revert it at one point because it was causing issues when creating views with large PMS symbology.  When you had investigated it earlier this year you said it was because the json was too large and potentially hitting the limit for the async call
https://devtopia.esri.com/WebGIS/arcgis-portal-app/issues/30286

Is this still likely to be an issue?  


From: Khaled Hassen <KHassen@esri.com> 
Sent: January 7, 2021 9:49 AM
To: Michael Miller <mmiller@esri.com>
Cc: Chris Fox <chris_fox@esri.com>; Chris Whitmore <cwhitmore@esri.com>; Paul Barker <PBarker@esri.com>; Jeremy Bartley <jbartley@esri.com>
Subject: RE: Time out on Add To Definition

Hi Jeremy/Paul/Chris,

It does seem we are using async=false when calling addToDefinition when creating the view in the UX.
We will need to switch to async=true similar to other admin API usage in the UX.

Can you pl. add this as an enhancement in the UX side.

Michael: until this is resolved, I think you can use the admin API directly to do this for such large addToDefinition call.

Thanks

Khaled

From: Michael Miller <mmiller@esri.com> 
Sent: Thursday, January 7, 2021 6:33 AM
To: Khaled Hassen <KHassen@esri.com>
Cc: Chris Fox <chris_fox@esri.com>
Subject: RE: Time out on Add To Definition

I was using the home app in AGOL, so I do not have control over this setting.  If it does not use aysnc, I think I should log an enhancement so it does.  

From: Khaled Hassen <KHassen@esri.com> 
Sent: Thursday, January 7, 2021 9:30 AM
To: Michael Miller <mmiller@esri.com>
Cc: Chris Fox <chris_fox@esri.com>
Subject: RE: Time out on Add To Definition

Hi Michael,

You would need to use async=true when apply such large admin API.
The async does not have time limit. Do you have any issues using addToDefinition with async=true?

Thanks

Khaled

From: Michael Miller <mmiller@esri.com> 
Sent: Thursday, January 7, 2021 1:56 AM
To: Khaled Hassen <KHassen@esri.com>
Cc: Chris Fox <chris_fox@esri.com>
Subject: FW: Time out on Add To Definition

Khaled,
                Is there a way to increase the time out of an AddToDefinition call of a hosted feature service?  We are attempting to create a view on a feature service with ~40 layers, each with a related table, so a total of 80 layers and 40 relationships.  This is timing out after a minute.  Is the bug with the client, that the client needs to chunk the add to definition calls when there is a large number of layers?

Here is an example where it timed out in 1 min where I just had a large number of layers and tables, no relationships.

**Comment by chris-fox** [comment]
@jmhauck, per the trail above, it seems we still need to chunk even with an async call. I think we go with chunks of 20 layers/tables and can have @MikeMillerGIS give that a go.

**Comment by chris-fox** [comment]
Verified
